File: Algorithm/Programmers/Queue/다리를지나는트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        
        Queue<Integer> wait = new LinkedList<>();
        Queue<Integer> bridge = new LinkedList<>();
        
        for(int t : truck_weights) wait.add(t);
        for(int i = 0; i<bridge_length; i++) bridge.add(0);

        int time = 0;

        while(!wait.isEmpty()){
            time++;
            bridge.poll();
            if(sumWeight(bridge)+wait.peek()<=weight){
                bridge.add(wait.poll());                
            }else{
                bridge.add(0);
            }
        }
          
        int cnt=0;
        for(int b : bridge) if(b>0) cnt++;
 
        int answer = time + (bridge_length-cnt) + cnt;
        
        return answer;
    }
    
    public int sumWeight(Queue<Integer> queue){
        int sum = 0;
        for(int q : queue){
            sum+=q;
        }
        return sum;
    }
}

File: Algorithm/Programmers/Queue/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let bridge = new Array(bridge_length).fill(0);
    let time = 0;
    
    while (bridge.length) {
        bridge.shift();
        time ++;

        if (truck_weights.length) {
            let sum = bridge.reduce((sum, current) => sum + current, 0);
          if (sum + truck_weights[0] <= weight) {
            bridge.push(truck_weights.shift());
          } else {
            bridge.push(0);
          }
        }
      }

  return time;
}
